public/scripts/logprobs.js

```diff
@@ -0,0 +1,455 @@
+        .sort(([, logA], [, logB]) => logB - logA)
+        .map(([text, log]) => {
+            const probability = Math.exp(log);
+            sum += probability;
+            return [text, probability, log];
+        });
+    candidates.push(['<others>', 1 - sum, 0]);
+
+    let matched = false;
+    for (const [token, probability, log] of candidates) {
+        const container = $('<button class="flex-container flexFlowColumn logprobs_top_candidate"></button>');
+
+        if (token === selectedToken) {
+            matched = true;
+            container.addClass('selected');
+        }
+
+        const tokenText = $('<span></span>').text(`${toVisibleWhitespace(token)}`);
+        const percentText = $('<span></span>').text(`${(probability * 100).toFixed(2)}%`);
+        container.append(tokenText, percentText);
+        container.attr('title', `logarithm: ${log}`);
+        addKeyboardProps(container);
+        if (token !== '<others>') {
+            container.click(() => onAlternativeClicked(state.selectedTokenLogprobs, token));
+        } else {
+            container.prop('disabled', true);
+        }
+        nodes.push(container);
+    }
+
+    // Highlight the <others> node if the selected token was not included in the
+    // top logprobs
+    if (!matched) {
+        nodes[nodes.length - 1].css('background-color', 'rgba(255, 0, 0, 0.1)');
+    }
+
+    view.append(nodes);
+}
+
+/**
+ * onSelectedTokenChanged is called when the user clicks on a token in the
+ * token output view. It updates the selected token state and re-renders the
+ * top logprobs view, or deselects the token if it was already selected.
+ * @param {TokenLogprobs} logprobs - logprob data for the selected token
+ * @param {Element} span - target span node that was clicked
+ */
+function onSelectedTokenChanged(logprobs, span) {
+    $('.logprobs_output_token.selected').removeClass('selected');
+    if (state.selectedTokenLogprobs === logprobs) {
+        state.selectedTokenLogprobs = null;
+    } else {
+        state.selectedTokenLogprobs = logprobs;
+        $(span).addClass('selected');
+    }
+    renderTopLogprobs();
+}
+
+/**
+ * onAlternativeClicked is called when the user clicks on an alternative token
+ * in the top logprobs view. It will create a new swipe message and prefill it
+ * with all text up to the selected token, followed by the chosen alternative.
+ * Then it requests a `continue` completion from the model with the new prompt.
+ * @param {TokenLogprobs} tokenLogprobs - logprob data for selected alternative
+ * @param {string} alternative - selected alternative token's text
+ */
+function onAlternativeClicked(tokenLogprobs, alternative) {
+    if (!checkGenerateReady()) {
+        return;
+    }
+
+    if (getGeneratingApi() === 'openai') {
+        return callPopup(`<h3>Feature unavailable</h3><p>Due to API limitations, rerolling a token is not supported with OpenAI. Try switching to a different API.</p>`, 'text');
+    }
```

Comment:
Technically this does work, but the experience is awful because Chat Completion models don't like to seamlessly continue a previous generation due to the forced `<im_start>assistant:` OpenAI inserts at the bottom of every message.  So even if you pick an alternative token and expect the AI to keep going from there, most of the time it restarts the entire message or sentence and screws up your message. Seemed better to just disable it.